File: TP1/particle.py
import math
from typing import Type

class Particle:
    def __init__(self, id,x,y,radio):
        self.id = id
        self.neighbours = []
        self.x = x
        self.y = y
        self.radio = radio
    
    def add_neighbour(self, particle):
        self.neighbours.append(particle)
        
    def __str__(self):
        return self.id
    
    def get_neighbours(self):
        return self.neighbours
    
    def get_coordenates(self):
        return self.x, self.y
    
    def is_neighbour(self,other_particle: Type[Particle],Rc):
        x1,y1 = other_particle.get_coordenates()
        return (math.sqrt((self.x - x1)**2 + (self.y - y1)**2) - self.radio*2) <= Rc
    
    #The same function to calculate the distance but taking into account circular grid
    def is_neighbour(self,other_particle: Type[Particle],Rc,correction_X,correction_y):
        x1,y1 = other_particle.get_coordenates()
        return (math.sqrt((self.x - (x1 - correction_X))**2 + (self.y - (y1 - correction_y))**2) - self.radio*2) <= Rc

File: TP1/src/main/java/main/Particle.java
package main;


import java.util.ArrayList;
import java.util.List;

public class Particle {
    private final int id;
    private final List<Particle> neighbours = new ArrayList<>();
    private final double x;
    private final double y;
    private final double radio;

    public Particle(id id, double radio ){
        this.id = id;
        this.radio = radio;
    }

    public void setPosition(double x,double y){
        this.x = x;
        this.y = y;
    }

    public void addNeighbour(Particle particle){
        if (!neighbours.contains(particle)) {
            neighbours.add(particle);
        }
    }

    public int getId() {
        return id;
    }

    public double getX(){
        return x;
    }

    public double getY(){
        return y;
    }

    public String getNeighbours() {
        StringBuilder n = new StringBuilder();
        for (Particle particle : neighbours){
            n.append(particle.id);
            n.append(", ");
        }
        return n.toString();
    }

    public boolean isNeighbour(Particle otherParticle, double Rc){
        double x1 = otherParticle.getX();
        double y1 = otherParticle.getY();
        return Math.sqrt(Math.pow(x - x1, 2) + Math.pow(y - y1, 2))-radio- otherParticle.radio <= Rc;
    }

    public boolean isNeighbour(Particle otherParticle,double Rc, double correctionX, double correctionY){
        double x1 = otherParticle.getX();
        double y1 = otherParticle.getY();
        return Math.sqrt(Math.pow((x-correctionX) - x1, 2) + Math.pow((y-correctionY)-y1,2))-radio- otherParticle.radio <= Rc;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Particle particle = (Particle) o;
        return id == particle.id;
    }
}
